cunumeric/array.py

```diff
@@ -3113,32 +3138,38 @@ def mean(
                 dtype = np.dtype(np.float64)
             else:
                 dtype = self.dtype
+
+        where_array = broadcast_where(where, self.shape)
         # Do the sum
         if out is not None and out.dtype == dtype:
             sum_array = self.sum(
-                axis=axis,
-                dtype=dtype,
-                out=out,
-                keepdims=keepdims,
+                axis=axis, out=out, keepdims=keepdims, where=where_array
             )
         else:
             sum_array = self.sum(
-                axis=axis,
-                dtype=dtype,
-                keepdims=keepdims,
+                axis=axis, keepdims=keepdims, where=where_array
```

Comment:
if `dtype` is passed to `sum`, it will create temporary array with the shape of self. So calling `astype` on the return of `sum` will safe us some memory.

Reply:
But by doing it this way, the summation for `np.int64` arrays happens using an intermediate of type `np.int64`, whereas the NumPy docs specifically note that "float64 intermediate and return values are used for integer inputs".

If this is important for the SLAC usecase, then I think the better solution would be to extend the template of `UnaryRedImplBody` with two "CODE" template parameters, one for the RHS (input array) type and one for the LHS (accumulator array) type. Then we can compute the sum of an `np.int64` array directly into an array of `np.float64` type, without converting the input to `np.float64`.